qutebrowser/utils/utils.py

```diff
@@ -548,6 +550,24 @@ def sanitize_filename(name: str,
 
     for bad_char in bad_chars:
         name = name.replace(bad_char, replacement)
+
+    # Truncate the filename if it's too long.
+    max_bytes = 255
+    root, ext = os.path.splitext(name)
+    root = root[:max_bytes - len(ext) - len(optional_suffix)]
+    excess = len(os.fsencode(root + ext + optional_suffix)) - max_bytes
+
+    while excess > 0 and root:
+        # Max 4 bytes per character is assumed.
+        # Integer division floors to -∞, not to 0.
+        root = root[:(-excess // 4)]
+        excess = len(os.fsencode(root + ext + optional_suffix)) - max_bytes
+
+    if not root:
+        raise ValueError("the extension or the optional suffix is too long")
```

Comment:
Hmm, I wonder what we should do in this case... Right now, if a website has a `Content-Disposition` header suggesting a filename like `'foo.' + 'a' * 256`, qutebrowser would crash. With what I've seen so far with download filenames, whatever edge case there is, people will reach it some day :laughing: 

So I wonder if we should just give up and return the original name at this point? It would then fail to download, but that's still better than crashing qutebrowser I suppose.

Reply:
Could we just trim the extension? If it's something like `'a' * 256` it won't hurt if it becomes `'a' * 255`, right? I propose this implementation: trim the root while possible, then trim the extension.

Reply:
Depending on the OS (and software handling it), changing the extension changes the "type" of a file. Then again, if you have a file a file with an extension with >10 chars or so, probably something went wrong anyways - so I'm fine with your approach.